packages/dds/matrix/src/test/memory/sharedMatrix.spec.ts

```diff
@@ -0,0 +1,186 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	type IMemoryTestObject,
+	benchmarkMemory,
+	isInPerformanceTestingMode,
+} from "@fluid-tools/benchmark";
+
+import { SharedMatrix } from "../../index.js";
+import { createLocalMatrix } from "../utils.js";
+
+describe("SharedMatrix memory usage", () => {
+	// The value to be set in the cells of the matrix.
+	const matrixValue = "cellValue";
+	// The test matrix's size will be 0*0, 10*10, 100*100, 1000*1000.
+	const matrixSizes = isInPerformanceTestingMode
+		? [10, 100, 1000]
+		: // When not measuring perf, use a single smaller data size so the tests run faster.
+			[0];
+
+	// The number of operations to perform on the matrix.
+	const operationCounts = isInPerformanceTestingMode
+		? [100, 1000]
+		: // When not measuring perf, use a single smaller data size so the tests run faster.
+			[10];
+
+	// IMPORTANT: variables scoped to the test suite are a big problem for memory-profiling tests
+	// because they won't be out of scope when we garbage-collect between runs of the same test,
+	// and that will skew measurements. Tests should allocate all the memory they need using local
+	// variables scoped to the test function itself, so several iterations of a given test can
+	// measure from the same baseline (as much as possible).
+
+	beforeEach(async () => {
+		// CAREFUL: usually beforeEach/afterEach hooks are used to initialize or interact with variables
+		// whose scope is the encompasing test suite, but that's a problem for memory-profiling tests.
+		// See the comment at the top of the test suite for more details.
+	});
+
+	afterEach(() => {
+		// CAREFUL: usually beforeEach/afterEach hooks are used to initialize or interact with variables
+		// whose scope is the encompasing test suite, but that's a problem for memory-profiling tests.
+		// See the comment at the top of the test suite for more details.
+	});
+
+	for (const matrixSize of matrixSizes) {
+		describe(`Size of ${matrixSize}*${matrixSize} SharedMatrix`, () => {
+			// Filter counts to ensure they do not exceed matrixSize
+			const validRemoveCounts = operationCounts.filter((count) => count <= matrixSize);
+
+			for (const count of operationCounts) {
+				// Test the memory usage of the SharedMatrix for inserting a column in the middle for a given number of times.
+				benchmarkMemory(
+					new (class implements IMemoryTestObject {
+						readonly title = `Insert a column in the middle ${count} times`;
+						readonly localMatrix: SharedMatrix = createLocalMatrix({
+							id: "testLocalMatrix",
+							size: matrixSize,
+							initialValue: matrixValue,
+						});
```

Comment:
I think this should go back to `beforeIteration`. If we keep it here, we're creating one matrix at the very beginning of the perf test, and then every time `run` is called to generate a new data point, it's working on a bigger matrix than the previous time it executed, so the final measurement could be skewed since every iteration is working on a different initial state.

Same for the ones below, and extra important for the tests that remove things; creating a matrix of size 100 and then trying to remove 10 rows/cols 20 times (assuming the perf test infra runs 20 iterations) will eventually fail, or at the very least produce incorrect results.

Reply:
I thought it only count the code block under run(). Updated.